File: typescript/src/Part1. Introduction To Functional Programming/4. Handling Erros Without Exceptions/option.ts
import { absurd } from 'fp-ts/lib/function';
import { List, foldRight, nil, cons } from '../3. Functional Data Structures/list';

class Some<A> {
  readonly _tag = "Some"
  constructor(readonly get: A) {}
}

class None {
  readonly _tag = "None"
}

export type Option<A> = Some<A> | None;

export const some = <A>(get: A): Option<A> => new Some(get);
export const none = (): Option<never> => new None(); 

interface Match<A, B> {
  None: () => B,
  Some: (v: A) => B
}

export const match = <A>(option: Option<A>) => <B>(m: Match<A, B>): B => {
  switch (option._tag) {
    case 'None':
      return m.None()
    case 'Some':
      return m.Some(option.get)
    default:
      return absurd(option)
  }
}

// Exercise 4-1
export const map = <A, B>(f: (a: A) => B, o: Option<A>): Option<B> => {
  return match(o)({
    'None': () => none(),
    'Some': (v) => some(f(v)),
  })
  
};

export const flatMap = <A, B>(f: (a: A) => Option<B>, o: Option<A>): Option<B> => 
  getOrElse(map(f, o), () => none());

export const getOrElse = <A>(o: Option<A>, d: () => A): A => {
  return match(o)({
    'None': () => d(),
    'Some': (v) => v
  })
};

export const orElse = <A>(o: Option<A>, ob: () => Option<A>): Option<A> => 
  getOrElse(map(some, o), ob);

export const filter = <A>(f: (a: A) => boolean) => (o: Option<A>): Option<A> => 
  flatMap(x => (f(x) ? some(x) : none()), o)

// Exercise 4-2
const mean = (ns: number[]): Option<number> => {
  const nsLen = ns.length;
  const nsSum = ns.reduce((acc, a) => acc + a, 0)

  return nsLen ? none() : some(nsSum / nsLen)
}

export const variance = (xs: number[]): Option<number> =>
  flatMap((m) => mean(xs.map(x => Math.pow(x - m, 2))), mean(xs))

// Exercise 4-3
export const map2 = <A, B, C>(f: (a: A, b: B) => C, a: Option<A>, b: Option<B>): Option<C> =>
  flatMap((x) => map((y) => f(x, y), b), a)
//  a._tag === 'None' || b._tag === 'None' ? makeNone() : makeSome(f(a.get, b.get))


// ** Exercise 4-4
export const sequence = <A>(a: List<Option<A>>): Option<List<A>> =>
  foldRight(a, some(nil()) as Option<List<A>>)((e, acc) => map2(cons, e, acc));

// Exercise 4-5
export const traverse = <A>(a: List<A>) => <B>(f: (a: A) => Option<B>): Option<List<B>> =>
  foldRight(a, some(nil()) as Option<List<B>>)((e, acc) => map2(cons, f(e), acc))

const id = <A>(x: A): A => x;
export const sequence2 = <A>(a: List<Option<A>>): Option<List<A>> => traverse(a)(id);

File: scala/fpinscala/src/main/scala/fpinscala/exercises/errorhandling/Option.scala
package fpinscala.exercises.errorhandling

import scala.{Option => _, Either => _, _}

sealed trait Option[+A] { self =>
  // Listing4-2, exercises 4-1
  def map[B](f: A => B): Option[B] =
    self match {
      case None      => None
      case Some(get) => Some(f(get))
    }

  def flatMap[B](f: A => Option[B]): Option[B] =
    map(f).getOrElse(None)

  def getOrElse[B >: A](default: => B): B =
    self match {
      case None      => default
      case Some(get) => get
    }

  def orElse[B >: A](ob: Option[B]): Option[B] =
    map(Some(_)).getOrElse(ob)

  def filter(f: A => Boolean): Option[A] =
    flatMap(a => if (f(a)) Some(a) else None)
}

case object None extends Option[Nothing]
case class Some[+A](get: A) extends Option[A]

object Option {
  // exercises 4-2
  def variance(xs: Seq[Double]): Option[Double] =
    mean(xs).flatMap { m =>
      mean(xs.map { x =>
        math.pow(x - m, 2)
      })
    }

  def mean(xs: Seq[Double]): Option[Double] =
    if (xs.isEmpty) None
    else Some(xs.sum / xs.length)

  // exercises 4-3
  def map2[A, B, C](a: Option[A], b: Option[B])(f: (A, B) => C): Option[C] =
    a.flatMap { x =>
      b.map { y =>
        (f(x, y))
      }
    }

  // @see 74p
  def map22[A, B, C](a: Option[A], b: Option[B])(f: (A, B) => C): Option[C] =
    for {
      x <- a
      y <- b
    } yield f(x, y)

  // TODO: exercises 4-4
  def sequence[A](a: List[Option[A]]): Option[List[A]] = ???

  // TODO: exercises 4-5
  def traverse[A, B](a: List[A])(f: A => Option[B]): Option[List[B]] = ???
}
